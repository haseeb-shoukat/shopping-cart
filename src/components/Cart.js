const Cart = ({ totalPrice, cartProducts }) => {
  return (
    <div className="overlay">
      <div className="cart">
        <div className="cart-heading">Your shopping cart</div>
        <div className="cart-products">
          {cartProducts.map((product, key) => {
            <div key={key} className="cart-product">
              <img
                className="product-image-c"
                src={`./images/clothing/${product.img}.png`}
                alt=""
              />
              <div className="product-name">{product.name}</div>
              <div className="product-price">
                ${product.totalPrice.toFixed(2)}
              </div>
              <input type="number" step="1" />
            </div>;
          })}
        </div>
        <div className="cart-total">Total price: ${totalPrice.toFixed(2)}</div>
        <button className="checkout-btn">Checkout</button>
        <button className="close-btn">Close</button>
      </div>
    </div>
  );
};

export default Cart;
