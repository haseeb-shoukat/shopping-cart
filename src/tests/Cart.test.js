import React from "react";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import userEvent from "@testing-library/user-event";
import Cart from "../components/Cart";

const Product = class {
  constructor(key, name, price, img, quantity = 0) {
    this.key = key;
    this.name = name;
    this.price = price;
    this.img = img;
    this.quantity = quantity;
    this.totalPrice = 0;
  }

  increment = jest.fn();
  decrement = jest.fn();
};

describe("Cart component", () => {
  const hideCart = jest.fn();
  const updateProducts = jest.fn();

  it("Renders component", () => {
    const { container } = render(
      <Cart
        hideCart={hideCart}
        totalPrice={0}
        cartProducts={[]}
        updateProducts={updateProducts}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it("Renders empty div when product with zero quantity passed in", () => {
    const fakeItem = new Product(
      1,
      "Matt Black Cap, Unbranded, Unisex",
      19.5,
      "cap"
    );

    const { container } = render(
      <Cart
        hideCart={hideCart}
        totalPrice={0}
        cartProducts={[fakeItem]}
        updateProducts={updateProducts}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it("Renders the item when product with non-zero quantity passed in", () => {
    const fakeItem = new Product(
      1,
      "Matt Black Cap, Unbranded, Unisex",
      19.5,
      "cap",
      1
    );

    render(
      <Cart
        hideCart={hideCart}
        totalPrice={0}
        cartProducts={[fakeItem]}
        updateProducts={updateProducts}
      />
    );

    expect(
      screen.getByText("Matt Black Cap, Unbranded, Unisex")
    ).toBeInTheDocument();
  });

  it("Runs appropriate function when incr/decr button is pressed", () => {
    const fakeItem = new Product(
      1,
      "Matt Black Cap, Unbranded, Unisex",
      19.5,
      "cap",
      1
    );

    render(
      <Cart
        hideCart={hideCart}
        totalPrice={0}
        cartProducts={[fakeItem]}
        updateProducts={updateProducts}
      />
    );

    const buttons = screen.getAllByRole("button");
    buttons.forEach((button) => {
      if (button.textContent === "Close" || button.textContent === "Checkout")
        return;
      userEvent.click(button);
    });
    expect(fakeItem.increment).toHaveBeenCalledTimes(1);
    expect(fakeItem.decrement).toHaveBeenCalledTimes(1);
    expect(updateProducts).toHaveBeenCalledTimes(2);
  });

  it("Runs hideCart function when Close button is clicked", () => {
    render(
      <Cart
        hideCart={hideCart}
        totalPrice={0}
        cartProducts={[]}
        updateProducts={updateProducts}
      />
    );

    const button = screen.getByRole("button", { name: "Close" });
    userEvent.click(button);
    expect(hideCart).toHaveBeenCalledTimes(1);
  });
});
