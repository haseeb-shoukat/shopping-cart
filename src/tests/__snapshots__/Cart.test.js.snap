// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart component Renders component 1`] = `
<div>
  <div
    class="overlay"
  >
    <div
      class="cart"
    >
      <div
        class="cart-heading"
      >
        Your shopping cart
      </div>
      <div
        class="cart-products"
      />
      <div
        class="cart-total"
      >
        Total price: $
        0.00
      </div>
      <button
        class="checkout-btn"
      >
        Checkout
      </button>
      <button
        class="close-btn"
      >
        Close
      </button>
    </div>
  </div>
</div>
`;

exports[`Cart component Renders empty div when product with zero quantity passed in 1`] = `
<div>
  <div
    class="overlay"
  >
    <div
      class="cart"
    >
      <div
        class="cart-heading"
      >
        Your shopping cart
      </div>
      <div
        class="cart-products"
      />
      <div
        class="cart-total"
      >
        Total price: $
        0.00
      </div>
      <button
        class="checkout-btn"
      >
        Checkout
      </button>
      <button
        class="close-btn"
      >
        Close
      </button>
    </div>
  </div>
</div>
`;
